python/mxnet/io/io.py

```diff
@@ -966,6 +988,164 @@ def creator(*args, **kwargs):
+    except Exception:
+        logging.info('pack_img error on file: %s' % img_path)
+        return
+
+def _validate_filenames(list_file, output_path):
+    """
+    Helper function to validate the file paths of
+    the input list file and output .rec file path.
+    Parameters
+    --------
+    list_file: input list file path
+    output_path: path to the output directory
+    """
+    if not os.path.isfile(list_file):
+        raise Exception("Input list file is invalid - \
+            1. Wrong filename or file path \n2. List file should be of format *.lst")
+    if not os.path.isdir(output_path):
+        raise Exception("Output path should be a directory where the \
+            rec files will be stored.")
+
+def _count_elem(iter):
+    """
+    Helper function to count the number of elements in
+    a generator.
+    Parameters
+    -----
+    iter: generator object
+    Returns
+    -----
+    count: total count of elements
+    """
+    cnt = itertools.count()
+    deque(zip(iter, cnt), 0)
+    return next(cnt)
+
+def im2rec(list_file, transformer, dataset_params, output_path):
+    """
+    API to convert the input image dataset into recordIO file format.
+
+    Parameters
+    ----------
+    list_file : str
+        List file name
+    transformer : Transforms object
+    dataset_params : dict object
+    output_path : str
+        output file path
+
+    Returns
+    -------
+    output_path : str
+        output path of the rec file
+
+    """
+    _validate_filenames(list_file, output_path)
+    fname = os.path.basename(list_file).split('.')[0]
+    shared_obj_mgr = SharedObjectManager()
+    shared_obj_mgr.start()
+
+    num_workers = dataset_params.get('num_workers', mp.cpu_count() - 1)
+    batch_size = dataset_params.get('batch_size', 4096)
+    parts = dataset_params.get('parts', 1)
+    color = dataset_params.get('color', 1)
+    encoding = dataset_params.get('encoding', '.jpg')
+    quality = dataset_params.get('quality', 95)
+
+    part_size = _count_elem(_read_list(list_file, 1)) / parts
+
+    data_batch_iter = _read_list(list_file, batch_size)
+    # A process-safe PriorityQueue
+    out_q = shared_obj_mgr.PriorityQueue(batch_size)
+    pool = mp.Pool(num_workers)
+    write_count = 0
+
+    for data_batch in data_batch_iter:
+        pool.map(partial(_read_worker, out_q, transformer,
+                         color, quality, encoding), data_batch)
+
+        while not out_q.empty():
+            if write_count % part_size == 0:
+            	out_rec = os.path.join(output_path, fname) + '-part' + str(write_count / part_size) + '.rec'
+            	out_idx = os.path.join(output_path, fname) + '-part' + str(write_count / part_size) + '.idx'
+            	record = recordio.MXIndexedRecordIO(os.path.join(out_idx), os.path.join(out_rec), 'w')
+            deq = out_q.get()
+            write_count += 1
+            _, s, item = deq
+            record.write_idx(item[0], s)
+        logging.info("batch done")
```

Comment:
provide progress to logging info, repeating 'batch done' is not user-friendly.

Reply:
i will update the log messages.